internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,281 @@
+	defer span.End()
+
+	iter := func(ctx context.Context) (storage.TupleIterator, error) {
+		return c.OpenFGADatastore.Read(ctx, store, tupleKey, options)
+	}
+
+	if options.Consistency.Preference == openfgav1.ConsistencyPreference_HIGHER_CONSISTENCY {
+		return iter(ctx)
+	}
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("%sr%s/%s", QueryCachePrefix, store, tuple.TupleKeyToString(tupleKey)))
+	return c.newCachedIterator(ctx, iter, b.String())
+}
+
+type iterFunc func(ctx context.Context) (storage.TupleIterator, error)
+
+func (c *CachedDatastore) newCachedIterator(ctx context.Context, dsIterFunc iterFunc, key string) (storage.TupleIterator, error) {
+	span := trace.SpanFromContext(ctx)
+	tuplesCacheTotalCounter.Inc()
+
+	cacheKey := cacheKeyFor(key)
+
+	if cacheKey == "" {
+		return dsIterFunc(ctx)
+	}
+
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter          storage.TupleIterator
+	tuples        []*openfgav1.Tuple
+	cacheKey      string
+	cache         storage.InMemoryCache[any]
+	ttl           time.Duration
+	maxResultSize int
+	closeOnce     sync.Once
+	wg            sync.WaitGroup
+	sf            singleflight.Group
+}
+
+func (c *cachedIterator) addToBuffer(t *openfgav1.Tuple) bool {
+	if c.tuples == nil {
+		return false
+	}
+	c.tuples = append(c.tuples, t)
+	if len(c.tuples) >= c.maxResultSize {
+		tuplesCacheDiscardCounter.Inc()
+		c.tuples = nil
+	}
+	return true
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) && !errors.Is(err, context.Canceled) && !errors.Is(err, context.DeadlineExceeded) {
+			c.tuples = nil // don't store results that are incomplete
+		}
+		return nil, err
+	}
+
+	c.addToBuffer(t)
+
+	return t, nil
+}
+
+// Stop see [Iterator.Stop].
+func (c *cachedIterator) Stop() {
+	c.closeOnce.Do(func() {
+		if c.tuples == nil {
+			c.iter.Stop()
+			return
+		}
+		// prevent goroutine if iterator was already consumed
+		ctx := context.Background()
+		if _, err := c.iter.Head(ctx); errors.Is(err, storage.ErrIteratorDone) {
+			c.flush()
+			c.iter.Stop()
+			return
+		}
+
+		c.wg.Add(1)
```

Comment:
Can you explain the purpose of the `wg`?  I don't see c.wg.Wait() anywhere?

Reply:
In the unit tests. If you have a solution for this please advice.

Reply:
Maybe at least add comment in line 186?

Reply:
Added a godoc on the struct field, it's only used for synchronization in testing.